dialogue-core/src/main/java/com/palantir/dialogue/core/ContentDecodingChannel.java

```diff
@@ -0,0 +1,236 @@
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.zip.GZIPInputStream;
+
+/**
+ * Adds support for transparently requesting and decoding <code>Content-Encoding: gzip</code> responses
+ * in a client agnostic way. Client implementations may choose to decompress GZIP data more efficiently
+ * if possible.
+ * This allows client implementations to avoid considering content-encoding in most cases, and
+ * sets a specific <code>Accept-Encoding</code> header to avoid potentially using an unexpected
+ * type based on client defaults (for example apache httpclient requests gzip and deflate by default).
+ * https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.3
+ * https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.11
+ */
+final class ContentDecodingChannel implements Channel {
+
+    private static final String ACCEPT_ENCODING = "accept-encoding";
+    private static final String CONTENT_ENCODING = "content-encoding";
+    private static final String CONTENT_LENGTH = "content-length";
+    private static final String GZIP = "gzip";
+
+    private final Channel delegate;
+
+    ContentDecodingChannel(Channel delegate) {
+        this.delegate = Preconditions.checkNotNull(delegate, "Channel is required");
+    }
+
+    @Override
+    public ListenableFuture<Response> execute(Endpoint endpoint, Request request) {
+        return Futures.transform(
+                delegate.execute(endpoint, acceptEncoding(request)),
+                ContentDecodingChannel::decompress,
+                MoreExecutors.directExecutor());
+    }
+
+    private static Request acceptEncoding(Request request) {
+        if (request.headerParams().containsKey(ACCEPT_ENCODING)) {
+            // Do not replace existing accept-encoding values
+            return request;
+        }
+        return Request.builder()
+                .from(request)
+                .putHeaderParams(ACCEPT_ENCODING, GZIP)
+                .build();
+    }
+
+    private static Response decompress(Response input) {
+        Optional<String> contentEncoding = input.getFirstHeader(CONTENT_ENCODING);
+        if (contentEncoding.isPresent() && GZIP.equals(contentEncoding.get())) {
+            return new ContentDecodingResponse(input);
+        }
+        return input;
+    }
+
+    private static final class ContentDecodingResponse implements Response {
+
+        private final Response delegate;
+        private final Map<String, List<String>> headers;
+        private final InputStream body;
+
+        ContentDecodingResponse(Response delegate) {
+            this.delegate = delegate;
+            this.headers = Maps.filterKeys(delegate.headers(), ContentDecodingResponse::allowHeader);
+            this.body = new DeferredGzipInputStream(delegate.body());
+        }
+
+        @Override
+        public InputStream body() {
+            if (body == null) {
```

Comment:
I don't really understand this little guy - if the `body` field is final and assigned in the constructor, how can we end up in the body of this `if`?

Reply:
oops, I don't think this conditional is necessary

Reply:
interesting that nothing about our static analysis tooling complained though!